### PR TITLE
Checkout testcases

### PR DESCRIPTION
**Pull request by hhsecond**
## Description
Test cases & name validity check for the branch name

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **[CONTRIBUTING](../CONTRIBUTING.rst)** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/hangar/repository.py [hunk 1]
```diff
@@ -651,6 +651,12 @@ def create_branch(self, branch_name, base_commit=None):
         bool
             if the operation was successful.
         '''
+        if not is_ascii_alnum(branch_name):
```
Comment (hhsecond) [inline src/hangar/repository.py]: I could move this to a separate PR if you want to keep this PR just for test cases

File: tests/conftest.py [hunk 2]
```diff
@@ -6,7 +6,11 @@
 
 from hangar import Repository
 import hangar
+from hangar.records import summarize
 
+
+# TODO: IMO, we should not have this cleanup and let tests catch issues
+# if there are any
```
Comment (rlizzo) [inline tests/conftest.py]: Can you expand on this? 

I believe it is actually required, since without explicitly cleaning up the past instances, when a new repository is initiated with the same python process, the old state would still be left in the singleton instance dicts. This was the fix which made tests actually work in the first place. Unless https://github.com/tensorwerk/hangar-py/pull/20 fixed this behavior (I'm not sure, haven't tested it myself) then I think it's still needed? Please correct me if I'm wrong though!

File: tests/test_checkout.py [hunk 3]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
-
-    def test_merge_multiple_checkouts(self):
-        pass
-
-    def test_merge_diverged(self):
-        pass
-
-    def test_merge_diverged_conflict(self):
-        pass
+    def test_operate_on_dataset_after_closing_old_co(self, repo, array5by7):
```
Comment (rlizzo) [inline tests/test_checkout.py]: style note: don't use abreviations in test case function names. I don't care if test case function names go over some arbitrary line length limit (90ish for the main code base) since they should be as descriptive as possible.

File: tests/test_checkout.py [hunk 4]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
-        assert np.allclose(r1_checkout.datasets['dset']['1'], r2_checkout.datasets['dset']['1'])
+        assert np.allclose(
+            r1_checkout.datasets['dset']['1'], r2_checkout.datasets['dset']['1'])
+        assert r1_checkout.metadata.get('init') == 'array5by7 added'
+        assert r2_checkout.metadata.get('init') == 'array5by7 added'
         r1_checkout.close()
         r2_checkout.close()
         w_checkout.close()
 
     def test_write_with_r_checkout(self, written_repo, array5by7):
```
Comment (rlizzo) [inline tests/test_checkout.py]: style note: no abbreviations in test case function names. 

NOTE: this applies for all test functions, no need for me to clutter the review with comments for every single one.

File: tests/test_checkout.py [hunk 5]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
-
-    def test_merge_diverged_conflict(self):
-        pass
+    def test_operate_on_dataset_after_closing_old_co(self, repo, array5by7):
+        co = repo.checkout(write=True)
+        dset = co.datasets.init_dataset('dset', prototype=array5by7)
+        co.commit()
+        co.close()
+        co = repo.checkout(write=True)
+        # TODO: what is the design here. IMO, the old instance should not execute
```
Comment (rlizzo) [inline tests/test_checkout.py]: **It absolutely SHOULD NOT WORK! Does it?? That's a huge bug if so!!**

File: tests/test_checkout.py [hunk 6]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
+        dset.add(array5by7, '1')
+        co.commit()
+        co.close()
+        # TODO: even after the close, I could still fetch the
```
Comment (rlizzo) [inline tests/test_checkout.py]: This should not work. #

File: tests/test_checkout.py [hunk 7]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
+
+    def test_invalid_name(self, repo):
+        with pytest.raises(ValueError):
+            repo.create_branch('dummy branch')
+
+    def test_new_branch_from_new_repo(self, repo):
+        branch = repo.create_branch('testbranch')
+        # TODO: it doesnt raise value error to the user
+        # instead return false and print out the value error
+        # Ideally, it should let users create new branch
```
Comment (rlizzo) [inline tests/test_checkout.py]: Nope. For now, I want users to explicitly say they want to create a branch. 

To better explain the reasoning here: I'm hard set on being explicit for branch operations because the model is slightly different from `Git`'s in the sense that there is no way to do a `write-enabled checkout` on a `detached HEAD commit`. All writes have to be performed on the `"tip" HEAD commit` of a branch. Without being explicit about the branch creation and checkout operations, this subtle difference might lead to confusion in day-to-day use.

File: tests/test_checkout.py [hunk 8]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
+        assert type(branch) is str
+        co = written_repo.checkout(write=True, branch_name=branch)
+        assert co._branch_name == branch
+        co.datasets['_dset']['1'] = array5by7
+        co.metadata.add('a', 'b')
+        co.commit()
+        co.close()
+        written_repo.merge('test merge', 'master', branch)
+        co = written_repo.checkout()
+        # TODO: It's good to give access to name of the branch from Readcheckout
```
Comment (rlizzo) [inline tests/test_checkout.py]: I agree in principle, but disagree in practice. Since a read checkout and write checkout can operate at the same time on the same repository, we should only provide the commit which the reader checkout is operating on, not the actual branch name. 

Imaging the following scenario where a reader checks out `branch "A"` which (at the time of checkout) has head commit `2`. 

```
                         Reader c/o 
0 ---------> 1
              \___________> 2
                         branch A
```

Then a writer check's out `branch "A"` as well

```
                           Reader c/o
0 ---------> 1             Writer c/o
              \___________> 2
                         branch A
```

And then (while the reader is still active) the writer makes `commit 3` on `branch "A"`

```
                        Reader c/o
0 ---------> 1                                next writer c/o
            \___________-> 2 _________________-> 3
                        branch A
```

If we provided the "branch name" to the reader checkout object, it would be defined when it was first created, but then could potentially change to a `null` value over the course of it's lifetime... this would be a crazy behavior for user defined code which might depend on a branch name.

If we only provide the commit hash, we can be sure that it will never change under any circumstances, and a user can actually write application code which depends on that value for whatever need they might have.

File: tests/test_checkout.py [hunk 9]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
+        assert (co.datasets['_dset']['1'] == array5by7).all()
+        assert co.metadata.get('a') == 'b'
+
+    def test_merge_without_closing_previous_checkout(self, written_repo, array5by7):
+        branch = written_repo.create_branch('testbranch')
+        co = written_repo.checkout(write=True, branch_name=branch)
+        co.datasets['_dset']['1'] = array5by7
+        co.commit()
+        # TODO: Need better error message
+        # lockHeld check at line #235 of diff.py never happens
```
Comment (rlizzo) [inline tests/test_checkout.py]: good catch. please file a bug and link it here.

File: tests/test_checkout.py [hunk 10]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
+            written_repo.merge('test merge', 'master', branch)
+
+    def test_name_conflict(self, written_repo, array5by7):
+        written_repo.create_branch('testbranch')
+        with pytest.raises(ValueError):
+            written_repo.create_branch('testbranch')
+
+    def test_delete_branch(self, written_repo):
+        with pytest.raises(NotImplementedError):
+            # TODO should raise "branch doesn't exist error"
```
Comment (rlizzo) [inline tests/test_checkout.py]: I haven't figured out how we should handle removing branches yet. the logic it a bit complicated because we need to figure out what data is referenced in other commits, and which can be safely removed from disk. please create an issue and link it here for this.

File: tests/test_checkout.py [hunk 11]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
+        written_repo.merge('commit message', 'master', branch2)
+        co = written_repo.checkout()
+        print(co.datasets['_dset']['1'])
+        print(co.metadata.get('a'))
+
+    def test_new_branch_from_where(self, written_repo, get_history, array5by7):
+        branch1 = written_repo.create_branch('testbranch1')
+        branch2 = written_repo.create_branch('testbranch2')
+        co1 = written_repo.checkout(write=True, branch_name=branch1)
+        # TODO: better way to get history: could be part of checkout itself
```
Comment (rlizzo) [inline tests/test_checkout.py]: agreed. please create issue and link it here.

File: tests/test_checkout.py [hunk 12]
```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
...
+        co.commit()
+        co.close()
+        co = written_repo.checkout(write=True, branch_name=branch2)
+        newarray = np.zeros_like(array5by7)
+        co.datasets['_dset']['1'] = newarray
+        co.metadata.add('a', 'c')
+        co.commit()
+        co.close()
+        written_repo.merge('commit message', 'master', branch1)
+        # TODO: This should raise a conflict coz of same key
```
Comment (rlizzo) [inline tests/test_checkout.py]: I'm not following, can you explain what the bug is a bit better?

File: tests/conftest.py [hunk 13]
```diff
@@ -69,3 +73,12 @@ def randomsizedarray():
     a = random.randint(2, 8)
     b = random.randint(2, 8)
     return np.random.random((a, b))
+
+
+@pytest.fixture()
+def get_history(scope='sesssion'):
```
Comment (rlizzo) [inline tests/conftest.py]: Please don't define this as a fixture function, it's a built in utility of the library, and should just be called directly from a repository object in any tests which need it via the method `repo.summary()` (optionally with `return_contents=True` for programatic evaluation.  No need to make test fixtures more complex then they already area :)

File: tests/test_checkout.py (hunk 12 above)
Comment (hhsecond) [reply tests/test_checkout.py]: in both branch1 and branch2, test case has `_dset` that has data on the key `1`. It is essentially a merge conflict while merging but what hangar currently does is overwrite.

File: tests/conftest.py (hunk 13 above)
Comment (hhsecond) [reply tests/conftest.py]: I was thinking we'll keep all the dependencies to fixtures and keep the `import`s minimal as possible (Just another way of test designs we used to follow). Let's do it your way

File: tests/test_checkout.py (hunk 12 above)
Comment (rlizzo) [reply tests/test_checkout.py]: damn ok. big bug here!

File: tests/conftest.py (hunk 2 above)
Comment (hhsecond) [reply tests/conftest.py]: Exactly, By cleaning up the past instances, we are avoiding a chance to find whether we have that issue fixed or not. I understand it is required if it is not fixed (for other tests to run) but my point is, we need to remember to remove this fixture once we confirm that particular issue is gone.

File: tests/test_checkout.py (hunk 5 above)
Comment (hhsecond) [reply tests/test_checkout.py]: created #36

File: tests/test_checkout.py (hunk 6 above)
Comment (hhsecond) [reply tests/test_checkout.py]: created #36

File: tests/test_checkout.py (hunk 12 above)
Comment (hhsecond) [reply tests/test_checkout.py]: created #37

File: tests/test_checkout.py (hunk 10 above)
Comment (hhsecond) [reply tests/test_checkout.py]: #38

File: tests/test_checkout.py (hunk 9 above)
Comment (hhsecond) [reply tests/test_checkout.py]: #39

File: tests/test_checkout.py (hunk 11 above)
Comment (hhsecond) [reply tests/test_checkout.py]: `repo.log(return_contents=True)` does this, right?

File: tests/test_checkout.py (hunk 8 above)
Comment (hhsecond) [reply tests/test_checkout.py]: This makes a lot of sense. Although, I have a slightly convoluted solution where we raise a warning if the head moved further when the user tries to fetch the branch name from a read checkout and return the branch name. If the head not moved, return the branch name without raising the warning. I am just thinking aloud and perhaps this could be a problem for another day.

File: tests/test_checkout.py (hunk 7 above)
Comment (hhsecond) [reply tests/test_checkout.py]: Correct me If I am wrong; What I understand from the context above is, on a new repository user should not be able to create a new branch until the first commit?

File: tests/conftest.py (hunk 2 above)
Comment (rlizzo) [reply tests/conftest.py]: Ok. Linking this discussion to #24 and #21 since they will decide how this is handled in the long term.

File: tests/test_checkout.py (hunk 8 above)
Comment (rlizzo) [reply tests/test_checkout.py]: So I double checked, and this information is already provided to each of the checkout objects in one of their underscore attributes:

https://github.com/tensorwerk/hangar-py/blob/065ae0735c630f627fb5f5d9cd634739408a5733/src/hangar/checkout.py#L42

As part of #32 I'll make a `commit` attribute available to the user (following the descriptor protocol to prevent a user setting this value to something else after construction)

I'll do the same for the `branch_name` attribute for a write-enabled checkout.

https://github.com/tensorwerk/hangar-py/blob/065ae0735c630f627fb5f5d9cd634739408a5733/src/hangar/checkout.py#L124

File: tests/test_checkout.py (hunk 11 above)
Comment (rlizzo) [reply tests/test_checkout.py]: yeah!

File: tests/test_checkout.py (hunk 7 above)
Comment (hhsecond) [reply tests/test_checkout.py]: @rlizzo I am good to merge the PR once this conversation can be resolved (either through a Fix or an issue that tracks this test case). Let me know your thoughts. As a first-time hangar user, I would expect hangar to let me create a branch on an empty repository

**Comment by rlizzo** [comment]
So I'm a bit confused as to the status of this? I think there's a lot of overlap with #32, and this branch doesn't seem like it's really kept up with the changes in master. 

Do you want to continue with this PR? Or include the relevant changes in #32 along with a most focused commit for the remainder?

**Comment by hhsecond** [comment]
The TODOs in this PR is not just related to #32. There are few other changes that we haven't looked into at all. So I would say, let's merge this and then along with rebasing #32, will sort out what's required there. Once #32 lands, we'll have a stable master on which I can do a clean up perhaps.

**Comment by rlizzo** [comment]
Ok sounds good. I'm just going to rerun the tests to make sure everything is OK, then will merge this and rebase #32.
